Validate URL strings in notificationPermissions() callback

beidson opened the pull request:
#### 9b41fc1d28820ad02376e962deb315ffd46e5ecf
<pre>
Validate URL strings in notificationPermissions() callback
<a href="https://bugs.webkit.org/show_bug.cgi?id=257987">https://bugs.webkit.org/show_bug.cgi?id=257987</a>
rdar://110594888

Reviewed by Brent Fulgham and Sihui Liu.

The idea here is the client passes back URL strings.
Which - we&apos;d expect - can parse into a URL and therefore into a security origin.

We&apos;ve seen clients pass back keys that don&apos;t meet these criteria.
In the mean time - while tracking those cases down - let&apos;s make WebKit validate the input.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PushAPI.mm:

Canonical link: <a href="https://commits.webkit.org/265092@main">https://commits.webkit.org/265092@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c85c3e1e6a92db93b8a507fd03264757b936906

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10061 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11465 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9822 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10011 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12472 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10775 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/11881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8899 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/11881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/9180 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9047 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/11881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9549 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7778 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/8717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2350 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/12943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9324 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PushAPI.mm:
```diff
@@ -248,6 +248,20 @@ - (void)dealloc
+TEST(PushAPI, notificationPermissionsDelegateInvalidResponse)
+{
+    auto configuration = createConfigurationWithNotificationsEnabled();
+    RetainPtr<FirePushEventDataStoreDelegate> delegate = adoptNS([FirePushEventDataStoreDelegate new]);
```
auto?

brentfulgham reviewed (approved):
Looks great! r=me

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -167,8 +167,12 @@ bool showNotification(const WebCore::NotificationData& data) final
-            result.set(origin.toString(), value.boolValue);
+            auto originString = WebCore::SecurityOriginData::fromURL(URL(key)).toString();
+            if (originString.isEmpty()) {
+                RELEASE_LOG(Push, "[WKWebsiteDataStoreDelegate notificationPermissionsForWebsiteDataStore:] returned a URL string that could not be parsed into a security origin. Skipping.");
```
This should probably RELEASE_LOG_ERROR if it's unexpected.

szewai reviewed (approved):
r=me

beidson committed:
Validate URL strings in notificationPermissions() callback
https://bugs.webkit.org/show_bug.cgi?id=257987
rdar://110594888

Reviewed by Brent Fulgham and Sihui Liu.

The idea here is the client passes back URL strings.
Which - we'd expect - can parse into a URL and therefore into a security origin.

We've seen clients pass back keys that don't meet these criteria.
In the mean time - while tracking those cases down - let's make WebKit validate the input.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PushAPI.mm:

Canonical link: https://commits.webkit.org/265092@main
